include/cutlass/numeric_conversion.h

```diff
@@ -2441,6 +2442,225 @@ struct FastNumericArrayConverter<int8_t, float, N, Round> {
+template <FloatRoundStyle Round>
+struct FastNumericArrayConverter<cutlass::half_t, uint8_t, 4, Round> {
+  using result_type = Array<cutlass::half_t, 4>;
+  using source_type = Array<uint8_t, 4>;
+  static FloatRoundStyle const round_style = Round;
+
+  CUTLASS_DEVICE
+  static result_type convert(source_type const &source) {
+    result_type result;
+    
+    uint32_t const* source_ptr = reinterpret_cast<uint32_t const*>(&source);
+    uint32_t* result_ptr = reinterpret_cast<uint32_t*>(&result);
+
+    result_ptr[0] = __byte_perm(source_ptr[0], 0x0, 0x4140);
+    result_ptr[1] = __byte_perm(source_ptr[0], 0x0, 0x4342);
+
+    asm volatile("add.u32 %0, %1, %2;\n" : "=r"(result_ptr[0]) : "r"(result_ptr[0]), "r"(0x66006600));
+    asm volatile("add.u32 %0, %1, %2;\n" : "=r"(result_ptr[1]) : "r"(result_ptr[1]), "r"(0x66006600));
+
+    asm volatile("sub.f16x2 %0, %1, %2;\n" : "=r"(result_ptr[0]) : "r"(result_ptr[0]), "r"(0x66006600));
+    asm volatile("sub.f16x2 %0, %1, %2;\n" : "=r"(result_ptr[1]) : "r"(result_ptr[1]), "r"(0x66006600));
+
+    return result;
+  }
+
+  CUTLASS_DEVICE
+  result_type operator()(source_type const &s) const { 
+    return convert(s);
+  }
+};
+
+/// Partial specialization for Array<cutlass::bfloat16_t, 4> <= Array<uint8_t, 4>
+template <FloatRoundStyle Round>
+struct FastNumericArrayConverter<cutlass::bfloat16_t, uint8_t, 4, Round> {
+  using result_type = Array<cutlass::bfloat16_t, 4>;
+  using source_type = Array<uint8_t, 4>;
+  static FloatRoundStyle const round_style = Round;
+
+  CUTLASS_DEVICE
+  static result_type convert(source_type const &source) {
+    result_type result;
+    Array<float, 4> tmp;
+
+    uint32_t const* source_ptr = reinterpret_cast<uint32_t const*>(&source);
+    uint32_t* tmp_ptr = reinterpret_cast<uint32_t*>(&tmp);
+
+    // __byte_perm simulates the add.u32 0x4B000000 to every u8 element of u8x4 source and stores 
+    // the result in tmp (without introducing extra cvt.u32.u8 instruction)
+    tmp_ptr[0] = __byte_perm(source_ptr[0], 0x4B000000, 0x7650);
+    tmp_ptr[1] = __byte_perm(source_ptr[0], 0x4B000000, 0x7651);
+    tmp_ptr[2] = __byte_perm(source_ptr[0], 0x4B000000, 0x7652);
+    tmp_ptr[3] = __byte_perm(source_ptr[0], 0x4B000000, 0x7653);
+
+    // Subtract the magic number 0x4B000000 from tmp in floating-point arithmetic to obtain final result
+    CUTLASS_PRAGMA_UNROLL
+    for (int i = 0; i < 4; ++i) {
+      tmp[i] = reinterpret_cast<float const &>(tmp_ptr[i]) - 8388608.f;
+    }
+
+    // on 3456x4096x8192 runs at 158 TFLOP/s
+    // Convert f32x2 to bf16x2 using `cvt.rn.b16x2.f32` instruction
+    NumericArrayConverter<cutlass::bfloat16_t, float, 4, Round> convert_f32_to_bf16;
+    result = convert_f32_to_bf16(tmp);
+
+    return result;
+  }
+
+  CUTLASS_DEVICE
+  result_type operator()(source_type const &s) const { 
+    return convert(s);
+  }
+};
+
+/// Partial specialization for Array<cutlass::bfloat16_t, 4> <= Array<int8_t, 4>
```

Comment:
What is the complete PTX / SASS that we end up generating here?

What we used when implementing this conversion in Triton is somewhat different, see openai/triton#1879.

Basically, we emit this PTX:

```
    "mov.b32 {s0, s1, s2, s3}, $2;               \n" // unpack
    "cvt.rn.f32.s8 f0, s0;                       \n" // no s8->bf16 pre-Hopper
    "cvt.rn.f32.s8 f1, s1;                       \n" // fi[0:15] is always 0
    "cvt.rn.f32.s8 f2, s2;                       \n" //
    "cvt.rn.f32.s8 f3, s3;                       \n" //
    "prmt.b32 $0, f0, f1, 0x7632;                \n" // f32->bf16 + pack
    "prmt.b32 $1, f2, f3, 0x7632;                \n" //
```

Are you expecting the conversion used here to be more efficient than what is in Triton?

Reply:
The PTX for the conversion here is as follows:
```
	prmt.b32 %r347,%r315,%r315,43682;
	prmt.b32 %r349,%r315,%r315,48051;
	add.s32 	%r693, %r343, 1262485504;
	mov.b32 	%f689, %r693;
	add.f32 	%f530, %f689, 0fCB400000;
	add.s32 	%r694, %r345, 1262485504;
	mov.b32 	%f690, %r694;
	add.f32 	%f529, %f690, 0fCB400000;
	add.s32 	%r695, %r347, 1262485504;
	mov.b32 	%f691, %r695;
	add.f32 	%f532, %f691, 0fCB400000;
	add.s32 	%r696, %r349, 1262485504;
	mov.b32 	%f692, %r696;
	add.f32 	%f531, %f692, 0fCB400000;
	cvt.rn.bf16x2.f32 %r351, %f529, %f530;
	cvt.rn.bf16x2.f32 %r352, %f531, %f532;
```

SASS is as follows:
```
                 PRMT R19, R4, 0xbbb3, R4 ;
                 IADD3 R19, R19, 0x4b400000, RZ ;
                 FADD R19, R19, -12582912 ;
                 F2FP.BF16.PACK_AB R57, R19, R10 ;
                 IADD3 R24, R24, 0x4b400000, RZ ;
                 FADD R24, R24, -12582912 ;
                 F2FP.BF16.PACK_AB R11, R7, R24 ;
```

This extracts and converts _two_ `s8`s packed into `s8x4` 32b registers and pack it into `bf16x2` 32b register. 

@gflegar, What is the SASS from the snippet you posted above?